R/Spectra-functions.R

```diff
@@ -176,3 +176,36 @@ addProcessing <- function(object, FUN, ...) {
     unsplit(res, f = f, drop = TRUE)
     ## unlist(res, recursive = FALSE, use.names = FALSE)
 }
+
+#' @export applyProcessing
+#'
+#' @rdname Spectra
+applyProcessing <- function(object, f = dataStorage(object),
+                            BPPARAM = bpparam(), ...) {
+    if (!length(object@processingQueue))
+        return(object)
+    if (isReadOnly(object@backend))
+        stop(class(object@backend), " is read-only. 'applyProcessing' works ",
+             "only with backends that support writing data.")
+    if (!is.factor(f))
+        f <- factor(f, levels = unique(f))
+    if (length(f) != length(object))
+        stop("length 'f' has to be equal to the length of 'object' (",
+             length(object), ")")
+    bknds <- bplapply(split(object@backend, f = f), function(z, queue) {
+        peaks(z) <- .apply_processing_queue(peaks(z), msLevel(z),
+                                            centroided(z), queue)
+        z
+    }, queue = object@processingQueue, BPPARAM = BPPARAM)
+    bknds <- backendMerge(bknds)
+    if (is.unsorted(f))
+        bknds <- bknds[order(unlist(split(seq_along(bknds), f),
```

Comment:
Just a quick question: Could we put this into `MsBackend::backendMerge` to avoid writing this construct multiple times in different places?

Reply:
Which part would you like to put in the backend? The `order(...)`? I'm a little hesitant, because the backends should not know anything about being splitted and how. In addition I would want to keep this in `Spectra` because then implementation is in our hands - otherwise this would have to be done in each `backendMerge` of any backend - assuming the developer actually does take care of re-ordering them properly.

Reply:
You are right that we can't assume that any backend developer will take care of this. From the design point of view the ordering is part of the backend (no backend operation should change the order). That's why I suggest to put it into `MsBackend::backendMerge` than every backend developer has just to call `callNextMethod`. But this will change `MsBackend` from an interface to a class (and will not force the developer of backends to implement `backendMerge`).

Currently we use this `order` construct just in `applyProcessing` and `setBackend`. So I think we could keep it that way.

Reply:
Totally agree with you @sgibb . Would be clean to have that in the backend, but on the other side, the `Spectra` does something here the `MsBackend` doesn't know about: it is splitting the `MsBackend` by file to process them in parallel. I would want to have this parallel processing functionality in `Spectra`, so we have control over it in a central place.

Eventually we might find someday a better solution.